# EDA


## 🔹 1. Introducción
Este análisis explora un dataset de ventas retail con el objetivo de entender su estructura, calidad de datos y preparar el diseño de un modelo estrella.

## 🔹 2. Carga de datos

In [1]:
import pandas as pd
from config import*
import sys


df = pd.read_csv(FILE_PATH,encoding='latin-1')

# vista inicial
df.head()



,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


## 🔹 3. Estructura del dataset

### 📌 Dimensiones y columnas

In [2]:
# dimensiones del data frame
print(f"dimension del dataframe :{df.shape}")

dimension del dataframe :(9994, 21)


In [3]:


# columnas
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='str')

In [4]:
# informacion general
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   str    
 2   Order Date     9994 non-null   str    
 3   Ship Date      9994 non-null   str    
 4   Ship Mode      9994 non-null   str    
 5   Customer ID    9994 non-null   str    
 6   Customer Name  9994 non-null   str    
 7   Segment        9994 non-null   str    
 8   Country        9994 non-null   str    
 9   City           9994 non-null   str    
 10  State          9994 non-null   str    
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   str    
 13  Product ID     9994 non-null   str    
 14  Category       9994 non-null   str    
 15  Sub-Category   9994 non-null   str    
 16  Product Name   9994 non-null   str    
 17  Sales          9994 non-null   float64
 18  Quantity       9994

### 🧠 Análisis:
*El dataset contiene 9994 registros y 21 columnas, lo que lo hace manejable para análisis en memoria con pandas.
*La mayoría de las columnas son de tipo object (string), lo cual es consistente con variables categóricas como cliente, producto y ubicación.
*Se identifican variables numéricas clave para el análisis de negocio:
    Sales (ventas)
    Profit (ganancia/pérdida)
    Discount (descuento aplicado)
    Quantity (cantidad vendida)
*Se detecta que las columnas de fecha (Order Date, Ship Date) están en formato string, lo que requerirá conversión a tipo datetime para análisis temporal.
*Existen múltiples columnas identificadoras (Order ID, Customer ID, Product ID), lo que sugiere una estructura relacional del dataset.

## 🔹 4. Calidad de datos

In [5]:
# 5. Duplicados
duplicados=df.duplicated().sum()
print(f"numero de registros duplicados: {duplicados}")

numero de registros duplicados: 0


👉 comentario: No se encontraron registros duplicados



## 📌 Valores nulos

In [6]:
df.isnull().sum()

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

👉 comentario:

No hay valores nulos

## 5. Análisis de granularidad

In [7]:
id_unicos=df["Order ID"].nunique()
numero_registros = df.shape[0]
print(f' El numero de Ordenes unicas es {id_unicos}, menor al numero de registros o filas {numero_registros}')

 El numero de Ordenes unicas es 5009, menor al numero de registros o filas 9994


In [8]:
df.groupby("Order ID")["Order Date"].nunique().value_counts()

Order Date
1    5009
Name: count, dtype: int64

👉 comentario:

Cada orden puede tener múltiples productos, lo que indica que el dataset está a nivel de línea de pedido.

### 🔹 6. Análisis de fechas

In [9]:
# Validar fechas 
df["Order Date"] = pd.to_datetime(df["Order Date"], format="%m/%d/%Y")
df["Ship Date"] = pd.to_datetime(df["Ship Date"], format="%m/%d/%Y")

(df["Ship Date"] < df["Order Date"]).sum()

np.int64(0)

👉 comentario:

No se encontraron inconsistencias temporales

## 🔹 7. Análisis de métricas

In [10]:
#  Estadísticas
df[["Sales", "Profit", "Discount"]].describe()

,Sales,Profit,Discount
count,9994.000000,9994.000000,9994.000000
mean,229.858001,28.656896,0.156203
std,623.245101,234.260108,0.206452
min,0.444000,-6599.978000,0.000000
25%,17.280000,1.728750,0.000000
50%,54.490000,8.666500,0.200000
75%,209.940000,29.364000,0.200000
max,22638.480000,8399.976000,0.800000


👉 comentario:

existen valores negativos en profit
descuentos altos (hasta 80%)
alta dispersión en ventas

## 🔹 8. Relación entre variables

In [11]:
# . Relación descuento vs profit
df[["Discount", "Profit"]].corr()

,Discount,Profit
Discount,1.000000,-0.219487
Profit,-0.219487,1.000000


👉 comentario:

Existe una relación negativa débil entre descuento y profit

### 🔹 9. Análisis de outliers

In [12]:
df[df["Profit"] < -1000][["Sales", "Profit", "Discount"]]

,Sales,Profit,Discount
27,3083.430,-1665.0522,0.5
165,8159.952,-1359.9920,0.4
683,7999.980,-3839.9904,0.5
1199,1088.792,-1850.9464,0.8
1803,4663.736,-1049.3406,0.2
2697,22638.480,-1811.0784,0.5
2846,762.594,-1143.8910,0.8
2928,1633.188,-1306.5504,0.7
3011,2549.985,-3399.9800,0.7
3151,1799.994,-2639.9912,0.7


👉 comentario:

Las mayores pérdidas están asociadas a altos descuentos

### 10. Cardinalidad y dimensiones

In [13]:
for col in ["Customer ID", "Product ID", "Category", "Region"]:
    print(col, df[col].nunique())

Customer ID 793
Product ID 1862
Category 3
Region 4


👉 comentario:

alta cardinalidad en clientes y productos
baja cardinalidad en categorías

## 11. Jerarquías

In [14]:
df[["Category", "Sub-Category"]].drop_duplicates()

,Category,Sub-Category
0,Furniture,Bookcases
1,Furniture,Chairs
2,Office Supplies,Labels
3,Furniture,Tables
4,Office Supplies,Storage
5,Furniture,Furnishings
6,Office Supplies,Art
7,Technology,Phones
8,Office Supplies,Binders
9,Office Supplies,Appliances


👉 comentario:

Cada Sub-Category pertenece a una sola Category

## 🔹 12. Conclusiones

🔹 1. Calidad de datos

✔ No se encontraron valores nulos
✔ No existen registros duplicados
✔ Las fechas son consistentes tras su conversión a tipo datetime
✔ Se identificó un problema de codificación del archivo (encoding), resuelto mediante lectura con latin-1

🔹 2. Problemas detectados
Columnas de fecha (Order Date, Ship Date) en formato string → requieren conversión a datetime
Postal Code en formato numérico → debe tratarse como string por ser un identificador
Row ID no aporta valor analítico → se excluirá en etapas posteriores

🔹 3. Estructura del dataset
El dataset presenta una granularidad a nivel de línea de pedido (order line)
Una misma orden (Order ID) puede contener múltiples productos
Esto lo hace adecuado como base para una tabla de hechos

🔹 4. Métricas clave
Sales → valor de la venta
Profit → ganancia o pérdida (puede ser negativa)
Discount → porcentaje de descuento aplicado (0 a 0.8)
Quantity → número de unidades vendidas

🔹 5. Insights relevantes
Existe una relación negativa débil entre Discount y Profit
Altos descuentos incrementan la probabilidad de generar pérdidas
Se observan pérdidas significativas en transacciones de alto valor con descuentos elevados
El dataset es adecuado para análisis de rentabilidad y desempeño comercial

🔹 6. Estructura para modelo estrella
✔ Tabla de hechos:
Sales
Profit
Discount
Quantity
✔ Dimensiones identificadas:
Cliente
Producto
Tiempo
Ubicación

🔹 7. Relaciones y jerarquías

Existe una dependencia funcional:

Sub-Category → Category
Esto define una jerarquía natural dentro de la dimensión producto

🔹 8. Conclusión general

El dataset presenta una estructura consistente, sin problemas críticos de calidad de datos, y con una granularidad adecuada para la construcción de un modelo estrella.
Se identificaron transformaciones necesarias que serán abordadas en la fase ETL, y se establecieron las bases para el diseño dimensional.